### PR TITLE
Add the ability to fake registry entries

### PR DESCRIPTION
**Pull request by ramidzkh**
Add the ability to fake blocks, items and potentially others with the client

Known issues:
- [X] ~~Custom blocks get sent as air (I have a WIP mildly-invasive fix)~~

### CODE REVIEW

**Comment by Restioson** [comment]
What does "fake blocks" mean?

**Comment by ramidzkh** [comment]
> What does "fake blocks, items" mean?

The server logically uses Items and Blocks, but data sent to the client uses something it can understand, such as replacing a modded item with a diamond.

File: src/main/java/xyz/nucleoid/plasmid/mixin/fake/IdListMixin.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        if (entry instanceof BlockState) {
+            BlockState state = (BlockState) entry;
+            Block block = state.getBlock();
+
+            if (block instanceof FakeBlock) {
+                return (T) ((FakeBlock<?>) block).getFaking(state);
+            }
+        }
+
+        if (entry instanceof ItemStack) {
```
Comment (gegy1000) [inline src/main/java/xyz/nucleoid/plasmid/mixin/fake/IdListMixin.java:33]: Is `IdList` ever used for `ItemStacks`?

**Comment by gegy1000** [comment]
Will the server try to sync the registry entries for the fake blocks/items?

File: src/main/java/xyz/nucleoid/plasmid/mixin/fake/IdListMixin.java (hunk 1 above)
Comment (ramidzkh) [reply src/main/java/xyz/nucleoid/plasmid/mixin/fake/IdListMixin.java:33]: No but other mods may have an IdList for ItemStacks, ¯\\_(ツ)_/¯

**Comment by ramidzkh** [comment]
I've tested this with a client with only Fabric API installed, and it works like a charm

File: src/main/java/xyz/nucleoid/plasmid/mixin/fake/IdListMixin.java (hunk 1 above)
Comment (gegy1000) [reply src/main/java/xyz/nucleoid/plasmid/mixin/fake/IdListMixin.java:33]: Would that even be possible? 😁 there’s no registry for ItemStacks like there is for BlockState

**Comment by gegy1000** [comment]
Not super familiar with how the handshake goes. I’m sure it’s fine if it’s working but I wonder if it is still syncing, and if that might have any unforeseen implications?

**Comment by oliviathevampire** [comment]
what would this allow for?

**Comment by Restioson** [comment]
AFAIK it would let you add a modded item presented to the player as a non modded item - e.g a gun presented as a hoe.

**Comment by ramidzkh** [comment]
image
The 'brick' is actually a custom block with a direction property. The fluid is probably not flowing since I rushed the testing fluid implementation. I forgot to include items in the screenshot, but they do work. Also persists save and such

**Pull request merged**